spotify.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	cfg := &oauth2.Config{
+		ClientID:     os.Getenv(ClientKey),
+		ClientSecret: os.Getenv(ClientSecret),
+		Scopes:       scopes,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  AuthURL,
+			TokenURL: TokenURL,
+		},
+	}
+
+	tr := &http.Transport{
+		TLSNextProto: map[string]func(authority string, c *tls.Conn) http.RoundTripper{},
+	}
+	ctx := context.WithValue(context.Background(), oauth2.HTTPClient, &http.Client{Transport: tr})
+
+	ts := cfg.TokenSource(ctx, &t)
+	token, err := ts.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	s.client = s.auth.NewClient(token)
+	close(s.Done)
+	s.ready = true
+
+	return s, nil
+}
+
+func (s *Spotify) Callback(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()
+	logf("callback invoked with code: %s and state: %s", q.Get("code"), q.Get("state"))
+
+	t, err := s.auth.Token(s.state, r)
+	if err != nil {
+		logf("unable to get token from Spotify: %s", err)
+		http.Error(w, "Couldn't get token from Spotify", http.StatusForbidden)
+		return
+	}
+
+	buf, err := json.Marshal(t)
+	if err != nil {
+		logf("unable to marshal token: %s", err)
+	}
+
+	logf("oauth2 token: %s", string(buf))
+
+	if sf := r.FormValue("state"); sf != s.state {
+		logf("state value mismatched (%s != %s)", sf, s.state)
+		http.NotFound(w, r)
+		return
+	}
+
+	logf("creating client with token")
+	s.client = s.auth.NewClient(t)
+	fmt.Fprint(w, "login process completed.")
+	s.ready = true
+	s.Done <- s.ready
+}
+
+func (s *Spotify) AuthURL() string {
+	return s.auth.AuthURL(s.state)
+}
+
+func (s *Spotify) Ready() bool {
+	return s.ready
+}
+
+func (s *Spotify) CurrentUser() (string, error) {
+	if !s.ready {
+		return "", ErrNotReady
+	}
+	user, err := s.client.CurrentUser()
+	if err != nil {
+		logf("unable to get current user")
+		return "", err
+	}
+
+	return string(user.ID), err
+}
+
+func (s *Spotify) Search(query []string) ([]string, error) {
+	if !s.ready {
+		return nil, ErrNotReady
+	}
+
+	var (
+		limit  = 1
+		offset = 0
+		opts   = spotify.Options{Limit: &limit, Offset: &offset}
+	)
+	var idx []string
+	for _, t := range query {
+		q :=
+			strings.TrimSpace(
+				strings.Replace(
+					// cleaning up query as majorly people use brackets to define genre
+					regexp.
+						MustCompile(`[\[(].*[])]`).
+						ReplaceAllString(t, ""),
+					"--",
+					"-",
+					-1),
+			)
+		logf("searching for %s", q)
+
+		results, err := s.client.SearchOpt(q, spotify.SearchTypeTrack, &opts)
+		if err != nil {
+			logf("error occurred while searching for %s: %s", q, err)
+			return nil, err
+		}
+
+		if results.Tracks != nil && len(results.Tracks.Tracks) > 0 {
+			logf("found tracks for %s, picking first track", q)
+			idx = append(idx, results.Tracks.Tracks[0].ID.String())
+		}
+	}
+
+	logf("found %d tracks for %d queries", len(idx), len(query))
+	return idx, nil
+}
+
+func breakIDs(s []spotify.ID, limit int) [][]spotify.ID {
+	l := float64(len(s))
+	f := math.Ceil(l / float64(limit))
+	r := int(f + math.Copysign(0.5, f))
```

Comment:
I do not understand this line.
Were you facing any issue with `r := int(f)`? Or is there any other case?

Reply:
Its just round functionality. But its unnecessary here as `limit` will always be positive.